surefire-api/src/main/java/org/apache/maven/surefire/api/stream/AbstractStreamDecoder.java

```diff
@@ -305,32 +313,24 @@ private String readString( @Nonnull final Memento memento, @Nonnegative final in
         memento.getDecoder().reset();
         final CharBuffer output = memento.getCharBuffer();
         ( (Buffer) output ).clear();
-        final ByteBuffer input = memento.getByteBuffer();
         final List<String> strings = new ArrayList<>();
         int countDecodedBytes = 0;
         for ( boolean endOfInput = false; !endOfInput; )
         {
-            final int bytesToRead = totalBytes - countDecodedBytes;
+            final int bytesToRead = totalBytes - countDecodedBytes; // our wish to read bytes as much as possible
             read( memento, bytesToRead );
-            int bytesToDecode = min( input.remaining(), bytesToRead );
+            final ByteBuffer input = memento.getByteBuffer();
+            int bytesToDecode = min( input.remaining(), bytesToRead ); // our guarantee of available bytes in buffer
             final boolean isLastChunk = bytesToDecode == bytesToRead;
             endOfInput = countDecodedBytes + bytesToDecode >= totalBytes;
-            do
-            {
-                boolean endOfChunk = output.remaining() >= bytesToRead;
-                boolean endOfOutput = isLastChunk && endOfChunk;
-                int readInputBytes = decodeString( memento.getDecoder(), input, output, bytesToDecode, endOfOutput,
-                    memento.getLine().getPositionByteBuffer() );
-                bytesToDecode -= readInputBytes;
-                countDecodedBytes += readInputBytes;
-            }
-            while ( isLastChunk && bytesToDecode > 0 && output.hasRemaining() );
-
-            if ( isLastChunk || !output.hasRemaining() )
-            {
-                strings.add( ( (Buffer) output ).flip().toString() );
-                ( (Buffer) output ).clear();
-            }
+            boolean endOfChunk = output.remaining() >= bytesToRead;
+            boolean endOfOutput = isLastChunk && endOfChunk;
+            int readInputBytes = decodeString( memento.getDecoder(), input, output, bytesToDecode, endOfOutput,
+                memento.getLine().getPositionByteBuffer() );
+            countDecodedBytes += readInputBytes;
+            strings.add( ( (Buffer) output ).flip().toString() );
+            ( (Buffer) output ).clear();
+            memento.getLine().setPositionByteBuffer( 0 );
```

Comment:
What's the reason of resetting line position here?

Will give this branch a try later. Maybe compare events dump before and after on single test runs, and just see if larger projects have no other issues. 

I think I already tried with removed do-while once, because it was suspicious at first look. But didn't want to change anything else in #518 , but my guess was that that wile condition is never true.

Reply:
@zoltanmeze  `memento.getLine()` holds the line which could not be properly decoded and it is eligible for printing as a corruption on the console.

Reply:
After properly decoded bytes, we clear the buffer and new bytes will be decoded. The corrupted has to reset to 0 and remaining is 0 too of course.

Reply:
If `MalformedFrameException` is caught, whole byte sequence is printed as a corrupted string starting by `memento.getLine().getPositionByteBuffer()` up to the buffer's position.

Reply:
Wouldn't [read](https://github.com/apache/maven-surefire/pull/519/files#diff-3d0df21c8b05874314acbad0ec8fc6778db9eebe5a08bba3bcff4f52872a4ce6R321) also reset this to 0 in every iteration starting from second?

Reply:
The memento holds the reference to the buffer too. And if I clean it up, I should reset the position via `memento.getLine().setPositionByteBuffer( 0 )`.